File: src/main/java/org/fscotto/blog/config/guice/GuiceVerticleFactory.java
package org.fscotto.blog.config.guice;

import com.google.inject.Injector;
import io.vertx.core.Verticle;
import io.vertx.core.spi.VerticleFactory;

public class GuiceVerticleFactory implements VerticleFactory {

  private static final String PREFIX = "it.plague";

  private final Injector injector;

  public GuiceVerticleFactory(Injector injector) {
    this.injector = injector;
  }

  @Override
  public String prefix() {
    return PREFIX;
  }

  @Override
  public Verticle createVerticle(String verticleName, ClassLoader classLoader) throws Exception {
    var verticle = VerticleFactory.removePrefix(verticleName);
    return (Verticle) injector.getInstance(classLoader.loadClass(verticle));
  }
}

File: src/main/kotlin/it/plague/blog/config/guice/GuiceVerticleFactory.kt
package it.plague.blog.config.guice

import com.google.inject.Injector
import io.vertx.core.Verticle
import io.vertx.core.spi.VerticleFactory

class GuiceVerticleFactory(private val injector: Injector) : VerticleFactory {

	companion object {
		private const val prefix = "it.plague"
	}

	override fun prefix(): String = prefix

	override fun createVerticle(verticleName: String, classLoader: ClassLoader): Verticle {
		val verticle = VerticleFactory.removePrefix(verticleName)
		return injector.getInstance(classLoader.loadClass(verticle)) as Verticle
	}
}
